common/src/main/java/com/microsoft/identity/common/internal/broker/SdmQrPinManager.kt

```diff
@@ -0,0 +1,104 @@
+ * - Whether camera consent should be suppressed.
+ * - The preferred authentication method for the device.
+ *
+ * It is initialized each time `GetPreferredAuthMethodMsalBrokerOperation` is called,
+ * since apps always contact the broker to get the preferred authentication method
+ * before starting the QR + PIN authentication flow.
+ *
+ * Note: If this object is not initialized, both flags default to `false`,
+ * meaning the device is not in SDM QR PIN mode.
+ */
+object SdmQrPinManager {
+
+    private const val TAG = "SdmQrPinManager"
+
+    /**
+     * This is the preferred authentication method for the device on SDM mode.
+     * This is set in the Authenticator app, using app configuration policies for managed Android Enterprise devices.
+     */
+    var preferredAuthMethod: String? = null
+
+    /**
+     * This is a flag to indicate if the device is on SDM QR PIN mode.
+     * If this is true, the device should prompt the user for camera consent.
+     */
+    var isDeviceOnSdmQrPinAuth = false
+        private set
+
+    /**
+     * This is a flag to indicate if the camera consent should be suppressed.
+     * If this is true, the device should not prompt the user for camera consent.
+     */
+    var isCameraConsentSuppressed = false
+        private set
+
+    /**
+     * This variable is used to store the last update time of the SDM QR PIN manager.
+     */
+    private var lastUpdateTime: Long = 0L
+
+    /**
+     * This method initializes the SDM QR PIN manager by checking the broker restrictions manager.
+     * It sets the [isDeviceOnSdmQrPinAuth], [preferredAuthMethod] and [isCameraConsentSuppressed] flags.
+     * This method is called each time `GetPreferredAuthMethodMsalBrokerOperation` is called,
+     * because the restrictions rarely change, only when a new policy is pushed to the device.
+     *
+     * @param brokerRestrictionsManager The broker restrictions manager to check the restrictions.
+     */
+    fun initializeSdmQrPinManager(brokerRestrictionsManager: IBrokerRestrictionsManager) {
+        Logger.info(TAG, "Initializing SDM QR PIN manager.")
+        val multiValueRequest = buildMultiValueRequest(
+            booleanKeys = setOf(SUPPRESS_CAMERA_CONSENT),
+            stringKeys = setOf(PREFERRED_AUTH_CONFIG)
+        )
+        val multiValues = brokerRestrictionsManager.getMultiValues(
+            brokerAppPackageName = BrokerData.prodMicrosoftAuthenticator.packageName,
+            bundleOfKeys = multiValueRequest
+        )
+        preferredAuthMethod = multiValues.getString(PREFERRED_AUTH_CONFIG)
+        if (preferredAuthMethod == PreferredAuthMethod.QR.value) {
+            isDeviceOnSdmQrPinAuth = true
+        }
+        isCameraConsentSuppressed = multiValues.getBoolean(SUPPRESS_CAMERA_CONSENT)
+        Logger.info(
+            TAG,
+            "preferredAuthMethod: $preferredAuthMethod," +
+                    " isCameraConsentSuppressed: $isCameraConsentSuppressed"
+        )
```

Comment:
Any reason you're doing reads in initialize itself?

Can we do it when getPreferredAuthMethod() and isDeviceOnSdmQrPinAuth is called?

Reply:
Those values ​​are unlikely to change since they are set by Intune policy, so I was thinking of avoiding calls to the broker and restrictions manager by simply reading all the values ​​at once and holding them for a while until the broker process is killed.

Reply:
Is GetPreferredAuthMethodMsalBrokerOperation called only once by caller?

This method is called by GetPreferredAuthMethodMsalBrokerOperation every time

Reply:
Is this always supposed to be called before CameraPermissionRequestHandler runs?

Reply:
Before making a token request,
1 Teams calls the GetPreferredAuthMethodMsalBrokerOperation method. 
2 If the preferred authentication method is QR+PIN, Teams displays a button for signing in using the camera. 
3 The user then logs in with QR+PIN.

Teams and other applications using this authentication method always call GetPreferredAuthMethodMsalBrokerOperation before making a token request with the camera. This is because they will only use QR+PIN authentication if the preferred method is QR+PIN, and to determine the method, they need to call GetPreferredAuthMethodMsalBrokerOperation.

Reply:
Agreed approach - I will read the value every time it is required, but I will initialize the object in the Android broker platform component.